### DISCUSSION TITLE
Windows Herd Slow

### DISCUSSION THREAD

**Discussion by lvismer** [beyondcode/herd-community#801] [Q&A]
Hi

I have a client who runs a Windows machine and using WAMP with a Laravel application (default Laravel 10) takes 6seconds and even more to load any page in the application.

I was hoping that Herd was our solution. But using Herd for windows the response on any page is still the same.  It is like the browser sits and waits (almost a resolving issue).

For others using Herd on Windows how is the response times for a Laravel application, is it snappy, ie 500ms or so?

Then is anyone experiencing the same on windows, and how did you solve the issue? I have the same issue on two different clients, with two different machines and setups. It makes Laravel almost unusable on Windows.

Any other pointers I can try. I'm wondering how I can test the DNS and mysql issue to make sure the bottleneck for waiting is not their. 

Thanks for any help and insights.

**Comment by mpociot** [comment]
Microsoft Defender seems to be a common performance issue. You can try and exclude the Herd folder (~/.config/Herd) as well as your site paths and see if this speeds things up. 

For me, the first request after starting Herd is always a bit slower and then it speeds up.

**Reply by lvismer** [reply]
@mpociot Thanks Marcel, will ask the client to try that. I know the client is using Bitdefender. I expected the first request to be slower but every request takes at least 6seconds, typically almost as if the google chrome cannot find the page, with the location spinner going.

**Reply by mauricekindermann** [reply]
I have tried this as well, didn't change anything. I still get 1500-3000+ms on windows, and 80-150ms on mac.

**Comment by RomkaLTU** [comment]
Experience same problem TTFB is 2s, I tried disabling defender, no dice. Unusable at this point, need to switch to WSL.

**Reply by RomkaLTU** [reply]
I created issue, this related to InertiaJS, maybe something todo with node, everythink is fine when rendering regular blade or serving API.

**Reply by mauricekindermann** [reply]
Yes unusable for me still, all this time later. Still working on a little mac laptop rather than my 3 screen windows beast. Crazy.

**Comment by pschilly** [comment]
Has there been any movement on this issue? I have seemingly randomly started having similar issues...  Was working fine up to this morning - now I get a `[2024-11-10 13:40:19] local.ERROR: Maximum execution time of 30 seconds exceeded {"exception":"[object] (Symfony\\Component\\ErrorHandler\\Error\\FatalError(code: 0): Maximum execution time of 30 seconds exceeded at `C:\\Users\\patri\\Herd\\heavy\\vendor\\symfony\\finder\\Iterator\\SortableIterator.php:51)` on every request (running Filament on Herd) - a standard install seems to work without issue though...

**Reply by pschilly** [reply]
For clarity - here is the full error stack: https://flareapp.io/share/bP9obBo5

Additionally - the exact same codebase that **was** working fine before - still is working fine via Laravel Sail on WSL2

I have tried the following:
- Complete fresh install of Herd
- Complete fresh install of the codebase from repo
- Complete fresh install of the codebase to an entirely different location on the computer (seperate physical drive)

Nothing seems to correct this issue :|

**Reply by mauricekindermann** [reply]
Yes I've tried reinstalling everything a few times, nothing fixes it. I have to imagine we are unlucky in some regard, surely it is not this slow on all PCs?

**Comment by ClaraLeigh** [comment]
Hey friends,

Here is some performance tips for windows (I'd PR this to docs but there is no github repo for it).

## File System Operations

> tl;dr reduce how much you are talking to the filesystem as much as possible. Your hdd speeds aren't the problem, php is

- Testing: Use DB transactions around your seeders (if you're seeding) to reduce file updates
- Debugbar: change the driver to redis from file, you'll see loads decrease by half using this alone
- Defender, disable it for the herd .config folder and your project folder

### `.env` file changes:

```
CACHE_STORE=file
SESSION_DRIVER=redis
```
Note: weirdly I notice faster speeds when using cache `file` over redis. I need to come back and investigate that quirk later. You may be different.

### PHP setting changes

Update both PHP and MySQL configs to tune for your system. 
I have a beefy system so here is my main changes.

First in herd settings change your memory to `-1` to disable the limit
Then in the `php.ini` - update values, don't add
```
realpath_cache_size = 16M
realpath_cache_ttl = 120
```

### MySQL Changes

`my.cnf` - mysql, I add all of these to the end. I have a 100+gb ram so decrease these values for your system 
```
innodb_buffer_pool_size=38G
innodb_log_file_size=4G
thread_cache_size=256
table_open_cache=20000
max_connections=300
innodb_io_capacity=10000
tmp_table_size=1G
max_heap_table_size=256M
innodb_log_buffer_size=128M
performance_schema=0
```

## Can we do more?

I suspect there is a lot more little tweaks that can likely do. I'll try to force myself to stay on windows for a little while as I know these things will annoy me until I find a way to fix it

**Reply by pschilly** [reply]
Well - this worked... Or seems to have.

Still **WAY** slower than it was originally - but it does load now...

**Reply by ClaraLeigh** [reply]
yeah honestly the problem is php in windows, see this: [https://bugs.php.net/bug.php?id=80695](https://bugs.php.net/bug.php?id=80695)

Looking around it looks like it's not the only windows quirk that slows things down

as said, I'll force myself to use windows for a month or so and I'll post back with anything else I find, oddly I like these sorts of challenges

**Reply by ClaraLeigh** [reply]
My experiment with windows has come to an end, never beat the slow first load which is what mainly has me back on my mac. It just doesn't seem to like creating lots of small files during the vite dev build. 

If someone ever cracks that nut, hmu

**Comment by pschilly** [comment]
Interesting... I'll give those steps a whirl and see what I find out.

**Reply by MGeurts** [reply]
To me, slow performance on Windows machines is always due to Microsoft Defender.
Excluding the Herd folder and my Development Folder, solved all performance issues.

Thanks ClaraLeigh!

**Comment by mauricekindermann** [comment]
Still unusable in 2025. My mac laptop is near instant, and windows, even with all of the above, has a minimum 2 second load per page.

**Reply by ClaraLeigh** [reply]
After the first page load, mine would be a lot faster. My bottleneck on windows was narrowed down to vite asset serving. 
I am back on my mac now but 2secs every time might be something like defender not respecting your rules (which is known to happen)

**Reply by mauricekindermann** [reply]
Empty controllers are around 2s. Livewire closer to 4s. 
![image](https://github.com/user-attachments/assets/39bafa11-e573-4231-8313-391d48058f80)

I've done all of the above several times over the last 18 months (perodically). 

I don't code on Windows anymore beacuse of this, my M.2 Macbook Air is around 80-120ms per request. Shame!

**Reply by avazquez-bit** [reply]
Windows is just an OS for playing... Only for playing... I develop in linux and had no problems with nothing about coding or performance, I started programing in windows, when my team invited to me to use linux I was not for sure, but now I pretty sure to neveer back to windows for any work... Just for playing, in this scenario is the best (I think) :)

**Reply by mauricekindermann** [reply]
> Windows is just an OS for playing... Only for playing... I develop in linux and had no problems with nothing about coding or performance, I started programing in windows, when my team invited to me to use linux I was not for sure, but now I pretty sure to neveer back to windows for any work... Just for playing, in this scenario is the best (I think) :)

Unhelpful argument and irrelevant to this thread. I've developed on windows for 15 years without speed issues, mostly raw PHP, but occasionally Zend, WordPress, and Cake. Laravel is the first PHP framework I've ever had run slow on windows, which is why I now can only use my Macbook for Laravel.

**Comment by SturmB** [comment]
I'm afraid I have to agree. I was hoping to abandon WSL and Sail for my Laravel dev work, but every request takes almost five seconds to complete. Back to WSL.

**Comment by chandra10207** [comment]
Adding Custom rules on windows firewall helped me:

Create a Firewall Rule:

Open Windows Defender Firewall with Advanced Security.

Go to Outbound Rules (or Inbound if incoming access matters).

Click New Rule > Choose Custom > Click Next.

Under Program, choose your app (or All Programs).

Under Protocol and Ports, click Next.

Under Scope, in the Remote IP address section:

Choose These IP addresses.

Add the resolved IP of the development.test. i.e 127.0.0.1 

Under Action, choose Allow the connection.

Apply it to the appropriate profiles (Domain, Private, Public).

Name the rule and click Finish.

**Reply by mauricekindermann** [reply]
Tried that just now, still 1500-3000ms per page request on Windows, and 80-150ms on mac.

**Reply by chandra10207** [reply]
Try to disable windows firewall for both private and public network and see whether its with firewall or something else.
Once disabled firewall its faster and then i add that rule and re-enabled the firewall.

**Reply by mauricekindermann** [reply]
I have tried that before, I've tried it again just now, same issue. So firewall is likely not the culprit. 

I've tried profiling with xDebug before, but herd injects so many files into each request it was impossible to teaase out what is going on. 

I am not a dumb dumb working on trash either. Running WAMP or Laragon with a plain PHP site is faster then my macbook running herd and laravel. It's only Windows + Herd + Laravel that is slow beyond use. 

I really hope I can work from windows one day, because it is a lot more enjoyable.

**Comment by mauricekindermann** [comment]
For what it's worth I did some more research and I didn't find anything useful at this point. 

- Disable Windows Defender or create exclusionsfor Herd (no improvement on my end)
- Try different  PHP config tweaks (nothing made a consquential difference for me)
- Switch back to Laragon, feasible, annoying, and faster but not lightning fast like Herd on Mac, so wouldn't bother here either
- Buy a new computer and switch operating systems - sadly and ironically this is the only solution that is guaranteed to work

I hope there is more to be done here one day.

**Comment by maxresnikoff** [comment]
My requests take around 70ms, but after enabling xdebug, my requests were taking 2s+.
Once disabled, it returned back to MacOS speeds. It might be worth checking if any PHP extensions in php.ini are the culprit for the high request overhead.

**Reply by mauricekindermann** [reply]
I could kiss you. THIS WAS IT! It was definitely very slow on launch, and I installed it to try and tease out what might be the bottleneck, but now it's much faster and disabling it was the fix I needed. Thankyou! 

My settings I had to comment out and restart:

```
# zend_extension=C:\Program Files\Herd\resources\app.asar.unpacked\resources\bin\xdebug\xdebug-8.3.dll
# xdebug.mode=profile
# xdebug.start_with_request=yes
# xdebug.start_upon_error=yes
# xdebug.output_dir=C:\Users\[username]\xdebug\profiles
# xdebug.profiler_output_name=cachegrind.out.%p
```

It's still not as fast as my Macbook, I get about 150ms-200ms total for a speciifc page, and windows gets about 300ms-400ms, but it's totally usable now and it wasn't before. Now I can at least upgrade my machine knowing it will make a difference :)

**Reply by mauricekindermann** [reply]
Clearly I need to upgrade my system regardless, because if you're seeing 70ms that's 5-6x what I'm seeing. Mind sharing your CPU/RAM/GPU specs? That would give everyone a new benchmark to aim towards for system requirements for fast develoment. Cheers mate :)

**Reply by maxresnikoff** [reply]
Glad to hear! 

Mind you, I am not running a Laravel project - it is a custom PHP REST API framework I have made, which might be why my requests might be faster than Laravel due to the lack of Laravel overhead(?)

Regardless, here are my specs (overkill for software dev, but is also my main gaming pc): 
- Intel(R) Core(TM) i(-14900K
- 64GB Ram
- NVIDIA GeForce RTX 4070 Ti SUPER
- NVME M.2 Hard drives (totalling 7TB)

**Reply by mauricekindermann** [reply]
Ahh ok, that explains it - pure PHP projects for me are basically instant. Thanks for sharing regardless!

**Comment by vorban** [comment]
With a default Herd installation, I've tried a bunch of things since I had a steady ~1sec response time on every request.
- Windows 11 with a decent gaming config.
- My database is a SQLite file on a SSD.
- Fresh Laravel 12 project with Livewire components, no starter kit, and an [icon library](https://blade-ui-kit.com/blade-icons?set=1)

The windows defender thing was barely noticeable, and I've tried multiple php.ini tweaks, with little success (perhaps 100ms total gain).

**The single most effective thing was `php artisan optimize`.**

I'm hovering between 50ms and 150ms for full pages and under 50ms for Livewire roundtrips.

Running the following commands does not drop the performances:
- `php artisan cache:clear`
- `php artisan route:clear`
- `php artisan view:clear`

I have little idea where that comes from, best guess is caching icon components from the library I use, since it plugs itself into the optimize pipeline.

**Reply by mauricekindermann** [reply]
Ha nice, that does make a decent difference. 

Ironically my problems started because of trying to xdebug-profile the slowness a year ago which I forgot I had enabled. 

I was pretty confident the issues stemmed from blade compontents / includes / nesting. This was confirmed by Caleb in his Livewire V4 talk last Laracon, which is why he's working on the code folding package to be released. 

But Optimize + clear cache/route/view is workable. I think I will make a custom command `optimize:windows` to run them.

**Reply by joaodemarco** [reply]
I was facing this same issue and after running the commands it got MUCH better and faster, tysm

**Comment by jellebreuer** [comment]
This fixed the issue for me, from 2s back to 20ms, 
put this in your .env file while developing
`INERTIA_SSR_ENABLED=false`

Im guessing the check for the ssr server is faster on osx/linux and it needs a full 2s timeout on windows

**Reply by ana-lisboa** [reply]
OMG Thank you so much! This made a world of a difference to me!
